### PR TITLE
Feature/migrations

### PR DESCRIPTION
**Pull request by germankrause**

### CODE REVIEW

**Review by msadovyi — APPROVED** [review]

File: packages/backend/src/common/tool/cipher.tool.ts [hunk 1]
```diff
@@ -0,0 +1,57 @@
+// migrations doesn't work without the reference
```
Comment (valcker) [inline packages/backend/src/common/tool/cipher.tool.ts:1]: migrations don't

File: packages/backend/src/common/tool/cipher.tool.ts [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+import { config } from 'node-config-ts';
+import crypto from 'crypto';
+
+const algorithm = 'aes-256-ctr';
+const secretKey = config.application.cryptoSecret;
+if (!secretKey || typeof secretKey !== 'string') {
+  throw new Error('config.application.cryptoSecret string is not set');
+}
+if (secretKey.length !== 32) {
+  throw new Error('config.application.cryptoSecret should have length 32');
```
Comment (valcker) [inline packages/backend/src/common/tool/cipher.tool.ts:13]: 'config.application.cryptoSecret should have length of 32 chars'

File: packages/backend/src/common/tool/index.ts [hunk 3]
```diff
@@ -12,3 +12,20 @@ export const randomEmail = (length = 10, domainLength = 3): string => {
 
 export const getObjectKey = (obj: { [key: string]: any }, path: string): any =>
   path.split('.').reduce((o, i) => o?.[i], obj);
+
```
Comment (valcker) [inline packages/backend/src/common/tool/index.ts:15]: Add a jsdoc comment with a short explanation of what it does + input/return values.

File: packages/backend/src/data/migrations/20210317182303-apidef-endpoint-unique.ts [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+
+export const up = async (db: Db): Promise<void> => {
+  const apiDefsCollection = await db.collection('apidefs');
+  const apiDefs: {
+    endpoint: string;
+  }[] = await apiDefsCollection.find().toArray();
+  const repeats = getArrayRepeats(apiDefs.map((apiDef) => apiDef.endpoint));
+  if (repeats.length) {
+    throw new Error(
+      `Cannot create index as apiDef.endpoint has repeats: ${repeats}`
```
Comment (valcker) [inline packages/backend/src/data/migrations/20210317182303-apidef-endpoint-unique.ts:14]: We are looking for non-unique apiDef.endpoint values, right? If yes, then we should provide an actionable error message:
`Cannot create an index as apiDef.endpoint "${apiDef.name} -> ${apiDef.endpoint}" is non unique and has duplicate values.`. Ideally, we should provide a list of apiDefs that we have to fix:
`Cannot create an index as apiDef.endpoint "${apiDef.endpoint}" is non-unique and is used in the following apiDefs: ${duplicateApiDefs.names}`.

File: packages/backend/src/common/tool/index.ts [hunk 5]
```diff
@@ -12,3 +12,20 @@ export const randomEmail = (length = 10, domainLength = 3): string => {
...
+export const getArrayRepeats = (arr: string[]): string[] => {
```
Comment (valcker) [inline packages/backend/src/common/tool/index.ts:16]: Knowing what it does, it should probably be called getArrayDuplicates

File: packages/backend/src/main.ts [hunk 6]
```diff
@@ -7,8 +7,16 @@ import AppModule from './modules/app.module';
 import Header from './common/enum/headers.enum';
 import { LoggerService } from './common/logger';
 import ValidationExceptionFilter from './common/error-handling/exception-filter';
+import { database, status } from 'migrate-mongo';
 
 const bootstrap = async (): Promise<void> => {
+  const { db } = await database.connect();
+  const migrations = await status(db);
+  if (!migrations?.length || migrations.pop()?.appliedAt === 'PENDING') {
+    console.error("Migrations are not applied, please run 'yarn migrate'");
```
Comment (valcker) [inline packages/backend/src/main.ts:16]: There are pending database migrations, please backup your database and run 'yarn migrate' to apply them.

File: packages/backend/migrate-mongo-config.js [hunk 7]
```diff
@@ -0,0 +1,18 @@
+const config = {
+  mongodb: {
+    url:
+      process.env.MONGODB_CONNECTION_STRING ||
```
Comment (valcker) [inline packages/backend/migrate-mongo-config.js:4]: Shall we take those values from node-config-ts?

**Review by valcker — APPROVED** [review]

File: packages/backend/migrate-mongo-config.js (hunk 7 above)
Comment (germankrause) [reply packages/backend/migrate-mongo-config.js:4]: yes, I've found a way to do it

**Pull request merged**